rubygems/rubygems
# Add poller to fetch WebAuthn OTP

## What was the end-user or developer problem that led to this PR?



Fixes: https://github.com/rubygems/rubygems.org/issues/3826
Discussed in the issue, we decided to rely on a polling solution for Safari users

RubyGems.org PR: https://github.com/rubygems/rubygems.org/pull/3873

## What is your fix for the problem, implemented in this PR?



If the user is using Safari, instead of redirecting the user to localhost after successful webauthn verification in the browser, [we'll show the success page](https://github.com/shopify/rubygems.org/blob/5cc347e37c8fcee61d9f463cba4f4c62fde46a6c/app/controllers/webauthn_verifications_controller.rb#L27) (since the auth on the browser was successful).

On the CLI, a poller is added that would poll `/api/v1/webauthn_verification/<token>/status` every 5 seconds.

Since we can't be sure which browser the user will be using (finding the default browser won't be fool proof since a user can paste the link into any browser), so we'll have a socket and poller thread open. 

Whichever thread terminates successfully with the otp will be used. This is done via a loop checking the status of the two threads.

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

## Testing

- Use this RubyGems.org branch https://github.com/rubygems/rubygems.org/pull/3873
- Have a Webauthn credential associated with your user.
- Run `RUBYGEMS_HOST=http://localhost:3000 ruby -Ilib exe/gem signin` and webauthn verify using Safari
